src/main/java/com/google/devtools/build/lib/skyframe/FilesystemValueChecker.java

```diff
@@ -441,7 +449,8 @@ private boolean artifactIsDirtyWithDirectSystemCalls(
       ImmutableSet<PathFragment> knownModifiedOutputFiles,
       boolean trustRemoteArtifacts,
       Map.Entry<? extends Artifact, FileArtifactValue> entry,
-      ModifiedOutputsReceiver modifiedOutputsReceiver) {
+      ModifiedOutputsReceiver modifiedOutputsReceiver,
+      long epochMilli) {
```

Comment:
What is the role of `FilesystemValueChecker` in this design? Is it merely an optimization to prevent spurious rewinding or is that necessary for correctness?

Reply:
it is an optimization to prevent rewinding.

Reply:
If that's solely an optimization, then I care significantly less about whether that seems unsound. One concern here is that e.g. outputs checking can be disabled for independent reasons (`--experimental_check_output_files`). Future features (e.g. imagine we added watchfs for outputs) could also break that in a way which is difficult to connect--that would be quite a leap for whoever implements such a feature to connect it with this behavior for remote execution.